File: src/runtime/hostCommand.ts
import { spawn } from "node:child_process";
import type { HostCommandSpec } from "../types.js";
import type { ArtifactStore } from "../store/artifacts.js";

const DEFAULT_MAX_OUTPUT_BYTES = 5 * 1024 * 1024;

export async function runHostCommand(spec: HostCommandSpec, cwd: string, artifactStore: ArtifactStore, inheritEnv: string[] = []): Promise<{
  status: "completed" | "failed";
  artifacts: Awaited<ReturnType<ArtifactStore["write"]>>[];
  data: {
    command: string[];
    exitCode: number | null;
    stdoutArtifactId: string;
    stderrArtifactId: string;
    stdoutTruncated: boolean;
    stderrTruncated: boolean;
    maxOutputBytes: number;
  };
}> {
  const maxOutputBytes = spec.maxOutputBytes ?? DEFAULT_MAX_OUTPUT_BYTES;
  const result = await runProcess(spec.argv[0]!, spec.argv.slice(1), cwd, spec.timeoutMs ?? 60_000, maxOutputBytes, buildEnv(inheritEnv));
  const stdout = await artifactStore.write(result.stdout, "text/plain", `hostCommand:${spec.id}:stdout`);
  const stderr = await artifactStore.write(result.stderr, "text/plain", `hostCommand:${spec.id}:stderr`);
  return {
    status: result.exitCode === 0 && !result.outputLimitExceeded ? "completed" : "failed",
    artifacts: [stdout, stderr],
    data: {
      command: spec.argv,
      exitCode: result.exitCode,
      stdoutArtifactId: stdout.id,
      stderrArtifactId: stderr.id,
      stdoutTruncated: result.stdoutTruncated,
      stderrTruncated: result.stderrTruncated,
      maxOutputBytes
    }
  };
}

function runProcess(
  command: string,
  args: string[],
  cwd: string,
  timeoutMs: number,
  maxOutputBytes: number,
  env: NodeJS.ProcessEnv
): Promise<{
  exitCode: number | null;
  stdout: string;
  stderr: string;
  stdoutTruncated: boolean;
  stderrTruncated: boolean;
  outputLimitExceeded: boolean;
}> {
  return new Promise((resolve) => {
    const child = spawn(command, args, {
      cwd,
      stdio: ["ignore", "pipe", "pipe"],
      env,
      shell: false,
      detached: true
    });
    let stdout: Buffer<ArrayBufferLike> = Buffer.alloc(0);
    let stderr: Buffer<ArrayBufferLike> = Buffer.alloc(0);
    let stdoutTruncated = false;
    let stderrTruncated = false;
    let outputLimitExceeded = false;
    const timer = setTimeout(() => {
      try {
        process.kill(-child.pid!, "SIGKILL");
      } catch {
        child.kill("SIGKILL");
      }
    }, timeoutMs);
    child.stdout.on("data", (chunk) => {
      const next = appendBounded(stdout, Buffer.from(chunk), maxOutputBytes);
      stdout = next.value;
      stdoutTruncated = stdoutTruncated || next.truncated;
      if (next.truncated) killForOutputLimit();
    });
    child.stderr.on("data", (chunk) => {
      const next = appendBounded(stderr, Buffer.from(chunk), maxOutputBytes);
      stderr = next.value;
      stderrTruncated = stderrTruncated || next.truncated;
      if (next.truncated) killForOutputLimit();
    });
    child.on("error", (error) => {
      clearTimeout(timer);
      const errorBuffer = Buffer.from(`\n${error.message}`, "utf8");
      const next = appendBounded(stderr, errorBuffer, maxOutputBytes);
      resolve({
        exitCode: null,
        stdout: stdout.toString("utf8"),
        stderr: next.value.toString("utf8"),
        stdoutTruncated,
        stderrTruncated: stderrTruncated || next.truncated,
        outputLimitExceeded
      });
    });
    child.on("close", (exitCode) => {
      clearTimeout(timer);
      resolve({
        exitCode,
        stdout: stdout.toString("utf8"),
        stderr: stderr.toString("utf8"),
        stdoutTruncated,
        stderrTruncated,
        outputLimitExceeded
      });
    });

    function killForOutputLimit(): void {
      if (outputLimitExceeded) return;
      outputLimitExceeded = true;
      try {
        process.kill(-child.pid!, "SIGKILL");
      } catch {
        child.kill("SIGKILL");
      }
    }
  });
}

function buildEnv(inherit: string[]): NodeJS.ProcessEnv {
  const env: NodeJS.ProcessEnv = { PATH: process.env.PATH ?? "" };
  for (const key of inherit) {
    if (key === "PATH") continue;
    const value = process.env[key];
    if (value !== undefined) env[key] = value;
  }
  return env;
}

function appendBounded(current: Buffer<ArrayBufferLike>, chunk: Buffer<ArrayBufferLike>, maxBytes: number): { value: Buffer<ArrayBufferLike>; truncated: boolean } {
  if (current.length >= maxBytes) return { value: current, truncated: chunk.length > 0 };
  const remaining = maxBytes - current.length;
  if (chunk.length <= remaining) return { value: Buffer.concat([current, chunk]), truncated: false };
  return { value: Buffer.concat([current, chunk.subarray(0, remaining)]), truncated: true };
}

File: src/host_command.rs
use crate::artifacts::ArtifactStore;
use crate::types::{ArtifactRecord, HostCommandSpec};
use anyhow::{Result, bail};
use serde_json::{Value, json};
use std::collections::BTreeMap;
use std::process::{Command, Stdio};

const DEFAULT_MAX_OUTPUT_BYTES: usize = 5 * 1024 * 1024;

#[derive(Debug, Clone)]
pub struct HostCommandResult {
    pub status: String,
    pub artifacts: Vec<ArtifactRecord>,
    pub data: Value,
}

pub fn run_host_command(
    spec: &HostCommandSpec,
    cwd: &std::path::Path,
    artifact_store: &ArtifactStore,
    inherit_env: &[String],
) -> Result<HostCommandResult> {
    let Some((program, args)) = spec.argv.split_first() else {
        bail!("host command argv must be non-empty");
    };
    let max_output_bytes = spec.max_output_bytes.unwrap_or(DEFAULT_MAX_OUTPUT_BYTES);
    let mut command = Command::new(program);
    command
        .args(args)
        .current_dir(cwd)
        .stdin(Stdio::null())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .env_clear();
    if let Some(path) = std::env::var_os("PATH") {
        command.env("PATH", path);
    }
    for key in inherit_env {
        if key == "PATH" {
            continue;
        }
        if let Some(value) = std::env::var_os(key) {
            command.env(key, value);
        }
    }
    let output = command.output()?;
    let (stdout, stdout_truncated) = bounded_output(&output.stdout, max_output_bytes);
    let (stderr, stderr_truncated) = bounded_output(&output.stderr, max_output_bytes);
    let stdout_record = artifact_store.write(
        &stdout,
        "text/plain",
        Some(&format!("hostCommand:{}:stdout", spec.id)),
    )?;
    let stderr_record = artifact_store.write(
        &stderr,
        "text/plain",
        Some(&format!("hostCommand:{}:stderr", spec.id)),
    )?;
    let exit_code = output.status.code();
    let output_limit_exceeded = stdout_truncated || stderr_truncated;
    Ok(HostCommandResult {
        status: if exit_code == Some(0) && !output_limit_exceeded {
            "completed".to_string()
        } else {
            "failed".to_string()
        },
        artifacts: vec![stdout_record.clone(), stderr_record.clone()],
        data: json!({
            "command": spec.argv,
            "exitCode": exit_code,
            "stdoutArtifactId": stdout_record.id,
            "stderrArtifactId": stderr_record.id,
            "stdoutTruncated": stdout_truncated,
            "stderrTruncated": stderr_truncated,
            "maxOutputBytes": max_output_bytes
        }),
    })
}

fn bounded_output(bytes: &[u8], max_bytes: usize) -> (Vec<u8>, bool) {
    if bytes.len() <= max_bytes {
        return (bytes.to_vec(), false);
    }
    (bytes[..max_bytes].to_vec(), true)
}

pub fn env_inherit_list(value: &Option<crate::types::EnvPolicy>) -> Vec<String> {
    value
        .as_ref()
        .map(|env| env.inherit.clone())
        .unwrap_or_default()
}

pub fn command_result_for_storage(status: &str, data: Value) -> Value {
    let mut object = BTreeMap::new();
    object.insert("status".to_string(), Value::String(status.to_string()));
    object.insert("data".to_string(), data);
    serde_json::to_value(object).unwrap_or(Value::Null)
}
